muxers/yamux/src/lib.rs

```diff
@@ -123,6 +140,8 @@ where
             return Poll::Ready(Ok(stream));
         }
 
+        self.inbound_stream_waker = Some(cx.waker().clone());
```

Comment:
The `poll` contract says that a waker needs to be registered in case `Poll::Pending` is returned. While perhaps this unconditional registration may be legal, I think it still violates expectations.

Reply:
This registration is not unconditional. It is conditional on the list being empty! The list no longer being empty is IMO a valid reason to wake the task that last polled here.

Reply:
Ah, are you saying that `self.poll_inner` will return `Poll::Pending`? That would be fine, then.

Reply:
OTOH: even if the inner stream does return `Poll::Pending`, it will have registered that same waker already, and it will thus wake it when new streams become available. What am I missing?

In general, the Rust async rules are thus: if you _create_ a `Poll::Pending`, then you’re responsible for waking a waker. Otherwise you just pass along poll results.

Reply:
> Ah, are you saying that `self.poll_inner` will return `Poll::Pending`? That would be fine, then.

Not quite.

This `poll` function essentially should we polled again in two different scenarios:

1. If the general poll-function pushed a new stream to the buffer.
2. If the socket has more bytes to read (that is `poll_inner`)

Registering this waker here isn't strictly necessary because we poll the `StreamMuxer` in `swarm::Connection` in a loop anyway. But I think it is more correct to still do this here because it showcases that there are two conditions on which the task should be polled again.

> OTOH: even if the inner stream does return `Poll::Pending`, it will have registered that same waker already, and it will thus wake it when new streams become available. What am I missing?

Like I said, it probably works without too because we will always be implicitly woken again, even if the task that calls the general `poll` consumes everything from the inner socket already and pushed them to the buffer.

Reply:
Making things fail due to a missed wake-up would require polling this task through `Arc<Mutex<...>>`, which I _hope_ nobody would ever consider doing, so I agree that this is harmless.

This discussion is yet another spotlight on how difficult it is to correctly reason about `poll` functions. This is exacerbated within libp2p by extending the concept — like is being done here — in non-trivial ways. If there actually was some surrounding task that did care about being woken for two different reasons, it would have to manufacture its own wakers (like FuturesUnordered does). But even that usage would be broken because the Waker destined for `poll_inbound` may be passed to `poll_inner` as well, overwriting a previously registered Waker that was destined for that purpose.

While debugging wake-up loops is painful, it is orders of magnitude easier than debugging missed wake-ups. May I suggest that we adhere to the policy that each async object offers exactly one `poll` function, with Future semantics, that drives exactly (i.e. only and completely) the state machine of that object? Interrogating the state machine should _not_ have `poll` semantics, because that can lead to confusing behaviour and bugs.

The following pattern is what I suggest:

```rust
if let Poll::Ready(x) = connection.poll(cx) {
    ...
    return x;
}
for sub in connection.inbound_streams.drain(..) {
    ...
}
```

Reply:
> May I suggest that we adhere to the policy that each async object offers exactly one `poll` function, with Future semantics, that drives exactly (i.e. only and completely) the state machine of that object? Interrogating the state machine should _not_ have `poll` semantics, because that can lead to confusing behaviour and bugs.

The design of this trait was inspired by the `AsyncWrite` design which also has multiple `poll_` functions that users need to drive. `Sink` is similar.

One problem with a single "poll" function design is that it puts more burden on implementations. For example, opening a new stream is not instantaneous, it may require negotiation of new credits with the other party. As such, a "new_outbound" function can only give you an ID or some other kind of handle for a new stream. This means every implementation needs to implement some kind of "stream ID" management. In contrast to that a `poll_new_outbound` function can just return `Pending` until the new stream is ready to be used.

Reply:
I never found Sink usable for anything I had to do, so its design carries little weight in my opinion.

Offering multiple “poll” functions to poll the same underlying thing has severe issues, as I argued above — and which you so far have not commented on. The example of a “new_outbound” function boils down to the choice of polling the machinery until the new connection is ready, ignoring everything else that happens in the meantime. This already requires the machinery to aggregate its output and let the poller inspect it, for which there is no reason a priori to offer a `poll`-shaped API. In particular, no `Context` is necessary to ask whether events have been emitted, which removes one prolific source of confusion inherent to Rust’s Task design.

So my solution to the `new_outbound` problem would be to offer a front-end Future that polls until the new connection is ready and leaves all other side-effects uninspected, to be dealt with by the caller afterwards.

Reply:
> So my solution to the `new_outbound` problem would be to offer a front-end Future that polls until the new connection is ready and leaves all other side-effects uninspected, to be dealt with by the caller afterwards.

This is basically the `poll_ready`& `start_send` API of `Sink` then, yes?

> Offering multiple “poll” functions to poll the same underlying thing has severe issues, as I argued above — and which you so far have not commented on.

`poll_outbound` is not the only issue. The problem of having to buffer streams is mostly because `yamux` doesn't allow us to backpressure the number of streams. The QUIC muxer on the other hand allows us to make progress on the connection itself without necessarily accepting new inbound streams. I am happy to change the API for something better but so far I've not found a solution where the caller (`swarm::Connection`) can explicitly signal to the muxer that it is now able to take more inbound streams.

We could move away from `poll_inbound` by having just a `pop_inbound` function. This however then requires more documentation on when the caller should call this function again if it ever returns `None`. At that stage, we are just re-inventing the wheel when we could also be using `Poll` and automatically wake the task when we know that there are new `inbound` streams available.